File: backend/schemas/scoring.py
"""Pydantic schemas for PriorityScore and DimensionScores."""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

DimensionState = Literal["green", "yellow", "red", "gray"]


class PriorityScore(BaseModel):
    """6-factor weighted priority score that drives the composite heatmap."""

    workload_id: str
    score: float = Field(ge=0, le=100, description="0-100, 1 decimal place")
    security_severity: float
    energy_waste: float
    cost_waste: float
    workflow_criticality: float
    environment_type: float
    self_healing_safety: float
    unavailable_factors: list[str] = Field(default_factory=list)
    detection_timestamp: datetime
    computed_at: datetime


class DimensionScore(BaseModel):
    """A single dimension's numeric score plus its state color."""

    score: float = Field(ge=0, le=100)
    state: DimensionState


class DimensionScores(BaseModel):
    """Per-dimension scores that drive the matrix heatmap."""

    workload_id: str
    security: DimensionScore
    energy: DimensionScore
    carbon: DimensionScore
    cost: DimensionScore
    performance: DimensionScore
    monitoring: DimensionScore

File: frontend/src/types/scoring.ts
// Mirrors backend/schemas/scoring.py and prediction.py.

export type DimensionState = 'green' | 'yellow' | 'red' | 'gray'

/** 6-factor weighted priority score that drives the composite heatmap. */
export interface PriorityScore {
  workload_id: string
  score: number // 0-100, 1 decimal place
  security_severity: number
  energy_waste: number
  cost_waste: number
  workflow_criticality: number
  environment_type: number
  self_healing_safety: number
  unavailable_factors: string[]
  detection_timestamp: string
  computed_at: string
}

/** A single dimension's numeric score plus its state color. */
export interface DimensionScore {
  score: number // 0-100
  state: DimensionState
}

/** Per-dimension scores that drive the matrix heatmap. */
export interface DimensionScores {
  workload_id: string
  security: DimensionScore
  energy: DimensionScore
  carbon: DimensionScore
  cost: DimensionScore
  performance: DimensionScore
  monitoring: DimensionScore
}

export type PredictionConfidence = 'low' | 'medium' | 'high'

/** Failure probability, timeline and contributing signals for a workload. */
export interface DowntimePrediction {
  workload_id: string
  probability: number // 0-100
  estimated_time_to_failure: string // "4h 30m" format
  primary_signal: string
  secondary_signal: string | null
  pattern_match: string | null
  confidence: PredictionConfidence
  risk_timeline: number[] // 12 points (hourly)
  recommended_preemptive_action: string | null // set when probability > 70%
}
